Optimization for CPU Causal Flash Attention (integrated into Qwen3)

DrJesseGlass opened the pull request:
This PR implements a complete CPU flash attention subsystem for candle, extending the original causal attention work with varlen support, interleaved KV cache, and fused zero-allocation decode kernels. Integrated into Qwen3 (full and quantized) and SmolLM3 (quantized with interlaced GGUF weights).
                                                                                                                                                                              
**Quantized Qwen3-0.6B Q4_K_M (MacBook Air, Apple Silicon,  22 prompt / 499 generated)):**
  | | Standard | CPU Flash | Delta |                                                                                                                                                                        
  |---|---|---|---|
  | Token/s | 26.42 | 36.89 | **+39.6%** |                                                                                                                                                                  
  | Steady-state RSS | ~800 MB | ~700 MB | **-12.5%**                                                                                                                                                   
                                                                                                                                                                                                            
  **Quantized SmolLM3-3B Q4_K_M (MacBook Air, Apple Silicon, 81 prompt / 499 generated):**                                                                                                                                                                                                            
  | | Standard | CPU Flash | Delta |                                                                                                                                                                      
  |---|---|---|---|                                                                                                                                                                                         
  | Token/s | 12.23 | 15.30 | **+25.1%** |                                                                                                                                                                  
  | Steady-state RSS | ~3450 MB | ~2020 MB | **-41.4%** |   
                                                                                                                                                                                                             
What changed                                                                                                                                                                                              
                                                                                                                                                                                                            
Attention module (candle-nn/src/attention/):                                                                                                                                                              
  - Shimmed old cpu_flash_attention (two functions changed to pub(crate)) and moved to modular attention/cpu_flash/ structure                                                                             
  - AttnMask enum with None, Causal { kv_offset }, Mask(tensor) for clean dispatch                                                                                                                          
  - B=1 kernels: standard.rs (explicit mask), causal.rs (loop-bound, ~50% less work)
  - B>1: auto-packs to varlen format — integrated from https://github.com/huggingface/candle/pull/3250                                                                                                      
  - Fused f32 decode kernels with prefetch hints (NEON/AVX), pre-allocated thread-local accumulators                                                                                                        
  - Interleaved KV decode kernel: K and V adjacent per position, single prefetch per KV step                                                                                                                
                                                                                                                                                                                                            
KV cache (candle-nn/src/kv_cache.rs):                                                                                                                                                                     
  - InterleavedKvCache: tensor-based interleaved cache for prefill                                                                                                                                          
  - RawInterleavedKvCache: pre-allocated Vec<f32> buffer, zero allocation on append                                                                                                                         
  - .detach() on ConcatKvCache and InterleavedKvCache to prevent BackpropOp memory leak — addresses the same issue as https://github.com/huggingface/candle/pull/3419                                     
  - In-place RoPE (rope_i_inplace) for fused decode without tensor ops                                                                                                                                      
                                                                                                                                                                                                            
Model integrations:                                                                                                                                                                                       
  - Qwen3 (qwen3.rs): CPU/GPU standard/flash dispatch, --use-flash-attn flag                                                                                                                                
  - Quantized Qwen3 (quantized_qwen3.rs): fused decode with raw cache, interleaved KV, conditional cache allocation                                                                                         
  - Quantized SmolLM3 (quantized_smollm3.rs): fused decode with interlaced GGUF weight support (skips reconstruction, uses rope_i for native element ordering)                                            
  - All flash paths guard B>1 — fall back to standard matmul for multi-batch

DrJesseGlass committed:
add flash attn to qwen3

DrJesseGlass committed:
feature flash-attn

DrJesseGlass committed:
merged main

DrJesseGlass committed:
flash generative true

DrJesseGlass committed:
no mask in cpu flash

DrJesseGlass committed:
add causal loop-bound optimization to cpu_flash_attention

DrJesseGlass committed:
attempt at hybrid

DrJesseGlass committed:
working tiling

DrJesseGlass committed:
working but poor performance tiled flash

DrJesseGlass committed:
factorized attention cpu flash

DrJesseGlass committed:
logging

DrJesseGlass committed:
causal cpu flash

DrJesseGlass committed:
depracation warning

DrJesseGlass committed:
depracation warning specific

DrJesseGlass committed:
formatting

DrJesseGlass committed:
interleaved will live in cpu_flash

DrJesseGlass committed:
clippy resolve

ivarflakstad reviewed (commented):
I'm waiting on https://github.com/huggingface/candle/pull/3250 but if it takes too long we'll incorporate these changes first :)

DrJesseGlass committed:
AttnMask: remove unnecessary lifetime

DrJesseGlass committed:
fix: flash-attn KV cache masking during decode

DrJesseGlass committed:
Merge branch 'main' into examples/qwen/add-flash

DrJesseGlass committed:
single batch CPU flash + integrate CPU varlen

DrJesseGlass committed:
fail B>1 my cpu flash

DrJesseGlass committed:
benchmark qwen3

DrJesseGlass committed:
clean up

DrJesseGlass committed:
varlen force

DrJesseGlass committed:
┌──────────────────┬──────────┬───────────┬────────┐
  │      Metric      │ Standard │ CPU Flash │ Delta  │
  ├──────────────────┼──────────┼───────────┼────────┤
  │ Token/s          │ 26.69    │ 27.36     │ +2.5%  │
  ├──────────────────┼──────────┼───────────┼────────┤
  │ Peak RSS         │ 1521 MB  │ 1411 MB   │ -7.2%  │
  ├──────────────────┼──────────┼───────────┼────────┤
  │ Steady-state RSS │ ~940 MB  │ ~830 MB   │ -11.7% │
  └──────────────────┴──────────┴───────────┴────────┘

DrJesseGlass committed:
dispatch logic smolLM3

DrJesseGlass committed:
interleaved cpu

DrJesseGlass committed:
interleaved smollm3

DrJesseGlass committed:
memory leak

DrJesseGlass committed:
back to exact exp

DrJesseGlass committed:
simplfy

DrJesseGlass committed:
cargo fmt

DrJesseGlass committed:
clippy

DrJesseGlass committed:
errors; fmt

DrJesseGlass committed:
cpu flash comment

DrJesseGlass committed:
remove bench script

DrJesseGlass commented:
Any updates or anything needed from me? I'm thinking to update this an remove the flag that you suggested.

DrJesseGlass committed:
Merge branch 'main' into examples/qwen/add-flash

DrJesseGlass committed:
merge main + resolve; default CPU flash (note that quantized varlen CPU flash not implemented) and remove --use-flash-attn CLI (note there is no path to standard attention for qwen3 and quantized qwen3)

DrJesseGlass commented:
I merged main and removed the use-flash-atten feature for CPUs -- note that as a result standard attention is never called for CPU when using the streamlined attention (except for quantized and batch_size > 1 because there is no quantized varlen CPU flash implementation yet). Cheers and GL.

DrJesseGlass committed:
debug cfg lazylock only

ivarflakstad committed:
Merge branch 'main' into examples/qwen/add-flash

ivarflakstad commented on candle-nn/src/attention/cpu_flash/mod.rs:
```diff
@@ -0,0 +1,167 @@
+static FORCE_VARLEN: LazyLock<bool> =
+    LazyLock::new(|| std::env::var("CANDLE_FLASH_FORCE_VARLEN").is_ok_and(|v| v == "1"));
+#[cfg(not(debug_assertions))]
+static FORCE_VARLEN: bool = false;
```
I assume you then use `--profile=release-with-debug`?
In just debug mode the generated binary would be so different that I suspect comparing paths would be futile.

I don't see any difference in output nor performance with the flag on or off though.

DrJesseGlass replied:
I used CANDLE_FLASH_FORCE_VARLEN during development to A/B the single-batch and varlen paths against each other -- running cargo run --release --profile-time with debug-assertions = true in a custom profile so the flag was live. Once I'd picked the B>1 threshold it stopped being useful.

ivarflakstad commented on candle-nn/src/kv_cache.rs:
```diff
@@ -723,26 +723,20 @@ impl ConcatKvCache {
-        // Update K cache using concatenation
+        // Detach inputs to break BackpropOp chain — KV caches are inference-only,
+        // no gradients flow through them. Without detach, Tensor::cat stores
+        // references to prior cache buffers in the op graph, causing O(N²) memory.
```
I'd prefer if the documentation was written by a human (I'm assuming you don't casually throw `U+00B2`, `U+002D`, and `U+2192` around).
Not because AI docs are inherently bad, in fact they can be quite good at documentation.
However these tools aren't great at brevity. And as we can see some lines below, the exact same concept is repeated.
In the general case (I know you're a good dev so doesn't apply) it is also a great early sign to me as a reviewer that what I am not wasting my time reviewing full slop.

Imo it's enough to simply state `// Detach inputs to break BackpropOp chain — KV caches are inference-only`. This is a comment for a fellow developer, as opposed to `///` , and so I expect that they either know exactly what this is referring to or gather from context that `.detach()` has some performance benefit.

DrJesseGlass replied:
Understood. Noted for future PRs. I will go over all the notes and documentation and simplify / write manually.

ivarflakstad reviewed (commented):
Thank you for your patience!
I'll have another look tomorrow 👍

DrJesseGlass committed:
remove unused cli

DrJesseGlass committed:
programatic commentary address

DrJesseGlass committed:
proposed comments for internal use

ivarflakstad commented on candle-nn/src/attention/cpu_flash/causal.rs:
```diff
@@ -0,0 +1,867 @@
+// f32 dot product
+
+#[inline(always)]
+fn dot_f32(a: &[f32], b: &[f32]) -> f32 {
```
Compilers are pretty amazing. This simple impl is faster than the target specific `CurrentCpu::vec_dot_f32` we have.

I did manage to improve the `CurrentCpu` to where it was slightly faster, but let's look into it later.

Here's updated version.
Made `const STEP` available and used it instead of using `fn n()`.
Changed iter.step_by to while loop.
Also added `#[inline(always)]` on all the `CurrentCpu` fns called here.

```rust
#[inline(always)]
pub(crate) unsafe fn vec_dot_f32(a_row: *const f32, b_row: *const f32, c: *mut f32, k: usize) {
    let mut sum = CurrentCpu::zero_array();
    let mut ax = CurrentCpu::zero_array();
    let mut ay = CurrentCpu::zero_array();

    let mut i = 0;
    while i + CurrentCpu::STEP < k {
        for j in 0..CurrentCpu::ARR {
            ax[j] = CurrentCpu::load(a_row.add(i + j * CurrentCpu::EPR));
            ay[j] = CurrentCpu::load(b_row.add(i + j * CurrentCpu::EPR));

            sum[j] = CurrentCpu::vec_fma(sum[j], ax[j], ay[j]);
        }
        i += CurrentCpu::STEP;
    }

    CurrentCpu::vec_reduce(sum, c);

    // leftovers
    while i < k {
        *c += *a_row.add(i) * (*b_row.add(i));
        i += 1;
    }
}
```

ivarflakstad commented on candle-nn/src/attention/cpu_flash/causal.rs:
```diff
@@ -0,0 +1,867 @@
+
+    FLASH_ATTN_POOL.install(|| {
+        out.par_chunks_mut(d)
+            .with_min_len(64)
```
Should look into this for the other variants as well

ivarflakstad reviewed (approved):
Great! Let's get this merged 🎉 

I'll add @michaelfeil as a co-author of the PR for the varlen work 👍

ivarflakstad commented:
I'd love for us to converge on a coherent design of the different cpu flash attn impls, without sacrificing performance ofc. Because with this merged we have 3 slightly different designs.
For example we have `dot_f32`, `vec_dot<T: ...>`, and candle's existing `f32::vec_dot` from `VecOps`.

DrJesseGlass commented:
I'll work on streamlining it to a single design next. Was trying for a small PR original mostly of dispatch logic but it grew due to varlen implementation conflict and then need to prove that non-varlen design was worth while.

DrJesseGlass commented:
@ivarflakstad I standardized the dot products behind a DotF32 trait in candle-nn/src/attention/cpu_flash/dot.rs. The f32 impl uses the compiler-vectorized body we both found fastest; the f16/bf16 impls accumulate in f32 (correctness fix over the old vec_dot::<T> which lost precision in T). I also factored the online softmax step into a shared helper at candle-nn/src/attention/cpu_flash/online_softmax.rs for cleaner kernel code #3513. 

Following on from that, I opened #3516 to bring the same compiler-vectorized body into candle-core::cpu::vec_dot_f32 itself, replacing the cfg-gated CurrentCpu intrinsics path. That makes conv1d / conv2d / convtranspose2d on the f32 CPU path use the same simpler implementation. The cfg gates and the unsafe pointer-arithmetic codepath go away; CurrentCpu and the f16/bf16/vec_sum paths are untouched. Conv2d microbench on M1 NEON shows it at least matching the intrinsics version (~4 percent faster on median, within noise) so the clear win there is maintenance. The design question you raised about whether a tweaked intrinsics version could win is still open and worth revisiting on AVX.
